### ISSUE TITLE
Make-Error Linux Mint 19

### ISSUE THREAD

**Issue by staxpost** [kylebarron/stata-png-fix#1]
I receive an error message using Linux Mint 19 XFCE (x64) with Stata15. I replicated the problem in a VM. First I installed Stata as usual. To get Stata running at all I have to manually download [this file](https://packages.ubuntu.com/xenial/amd64/libpng12-0/download) and install ist. Then I can run Stata, but the icons are missing.
I downloaded all files from github and made the required changes to **stata-png-fixed.sh**. I changed my folder to "/usr/local/stata15" and version to 15. Then I run 
`bash compile-stata-png-fix.sh`
This is the terminal output I receive:
```
hans@hans-VirtualBox:~/Downloads/stata-png-fix$ bash compile-stata-png-fix.sh
--2018-07-04 14:01:31--  http://downloads.sourceforge.net/project/libpng/zlib/1.2.3/zlib-1.2.3.tar.gz
Resolving downloads.sourceforge.net (downloads.sourceforge.net)... 216.105.38.13
Connecting to downloads.sourceforge.net (downloads.sourceforge.net)|216.105.38.13|:80... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://kent.dl.sourceforge.net/project/libpng/zlib/1.2.3/zlib-1.2.3.tar.gz [following]
--2018-07-04 14:01:32--  https://kent.dl.sourceforge.net/project/libpng/zlib/1.2.3/zlib-1.2.3.tar.gz
Resolving kent.dl.sourceforge.net (kent.dl.sourceforge.net)... 212.219.56.185
Connecting to kent.dl.sourceforge.net (kent.dl.sourceforge.net)|212.219.56.185|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 496597 (485K) [application/x-gzip]
Saving to: ‘zlib-1.2.3.tar.gz’

zlib-1.2.3.tar.gz   100%[===================>] 484,96K  1,09MB/s    in 0,4s    

2018-07-04 14:01:32 (1,09 MB/s) - ‘zlib-1.2.3.tar.gz’ saved [496597/496597]

--2018-07-04 14:01:32--  http://downloads.sourceforge.net/project/libpng/libpng16/older-releases/1.6.2/libpng-1.6.2.tar.gz
Resolving downloads.sourceforge.net (downloads.sourceforge.net)... 216.105.38.13
Connecting to downloads.sourceforge.net (downloads.sourceforge.net)|216.105.38.13|:80... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://netix.dl.sourceforge.net/project/libpng/libpng16/older-releases/1.6.2/libpng-1.6.2.tar.gz [following]
--2018-07-04 14:01:33--  https://netix.dl.sourceforge.net/project/libpng/libpng16/older-releases/1.6.2/libpng-1.6.2.tar.gz
Resolving netix.dl.sourceforge.net (netix.dl.sourceforge.net)... 87.121.121.2
Connecting to netix.dl.sourceforge.net (netix.dl.sourceforge.net)|87.121.121.2|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1256016 (1,2M) [application/x-gzip]
Saving to: ‘libpng-1.6.2.tar.gz’

libpng-1.6.2.tar.gz 100%[===================>]   1,20M  1,16MB/s    in 1,0s    

2018-07-04 14:01:34 (1,16 MB/s) - ‘libpng-1.6.2.tar.gz’ saved [1256016/1256016]

Checking for gcc...
Building static library libz.a version 1.2.3 with gcc.
Checking for unistd.h... No.
Checking whether to use vs[n]printf() or s[n]printf()... using s[n]printf()
Checking for snprintf() in stdio.h... No.
  WARNING: snprintf() not found, falling back to sprintf(). zlib
  can build but will be open to possible buffer-overflow security
  vulnerabilities.
Checking for return value of sprintf()... No.
  WARNING: apparently sprintf() does not return a value. zlib
  can build but will be open to possible string-format security
  vulnerabilities.
Checking for errno.h... No.
Checking for mmap support... No.
gcc -fPIC -DNO_snprintf -DHAS_sprintf_void -DNO_ERRNO_H   -c -o example.o example.c
example.c:8:10: fatal error: stdio.h: No such file or directory
 #include <stdio.h>
          ^~~~~~~~~
compilation terminated.
<builtin>: recipe for target 'example.o' failed
make: *** [example.o] Error 1
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking for gcc... gcc
checking whether the C compiler works... no
configure: error: in `/tmp/tmp.4VE0rIgi6w/libpng-1.6.2':
configure: error: C compiler cannot create executables
See `config.log' for more details
compile-stata-png-fix.sh: line 22: stata-png.patch: No such file or directory
make: *** No targets specified and no makefile found.  Stop.
install: cannot create regular file '/home/hans/local/stata-png-fix/stata-png-fixed.sh': No such file or directory
ln: failed to create symbolic link '/home/hans/local/bin/xstata': No such file or directory
ln: failed to create symbolic link '/home/hans/local/bin/xstata-se': No such file or directory
ln: failed to create symbolic link '/home/hans/local/bin/xstata-mp': No such file or directory
ln: failed to create symbolic link '/home/hans/local/bin/xstata-sm': No such file or directory

```

Finally I attach you the config.log that was created in the process.
[config.log](https://github.com/kylebarron/stata-png-fix/files/2163238/config.log)

**Comment by kylebarron** [comment]
It sounds like your C compiler isn't working. Did you install gcc on the VM?

**Comment by staxpost** [comment]
Well it was not installed but a "pure" Mint 19 system (I first encountered the problem at my regular laptop). I installed gcc now and I think it did change the error message, but a problem persists (see below). Is there any list of dependencies that I could try to install?

```
hans@hans-VirtualBox:~/Downloads/stata-png-fix$ bash compile-stata-png-fix.sh
--2018-07-04 18:23:28--  http://downloads.sourceforge.net/project/libpng/zlib/1.2.3/zlib-1.2.3.tar.gz
Resolving downloads.sourceforge.net (downloads.sourceforge.net)... 216.105.38.13
Connecting to downloads.sourceforge.net (downloads.sourceforge.net)|216.105.38.13|:80... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://datapacket.dl.sourceforge.net/project/libpng/zlib/1.2.3/zlib-1.2.3.tar.gz [following]
--2018-07-04 18:23:29--  https://datapacket.dl.sourceforge.net/project/libpng/zlib/1.2.3/zlib-1.2.3.tar.gz
Resolving datapacket.dl.sourceforge.net (datapacket.dl.sourceforge.net)... 185.152.64.70
Connecting to datapacket.dl.sourceforge.net (datapacket.dl.sourceforge.net)|185.152.64.70|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 496597 (485K) [application/x-gzip]
Saving to: ‘zlib-1.2.3.tar.gz’

zlib-1.2.3.tar.gz   100%[===================>] 484,96K   884KB/s    in 0,5s    

2018-07-04 18:23:30 (884 KB/s) - ‘zlib-1.2.3.tar.gz’ saved [496597/496597]

--2018-07-04 18:23:30--  http://downloads.sourceforge.net/project/libpng/libpng16/older-releases/1.6.2/libpng-1.6.2.tar.gz
Resolving downloads.sourceforge.net (downloads.sourceforge.net)... 216.105.38.13
Connecting to downloads.sourceforge.net (downloads.sourceforge.net)|216.105.38.13|:80... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://kent.dl.sourceforge.net/project/libpng/libpng16/older-releases/1.6.2/libpng-1.6.2.tar.gz [following]
--2018-07-04 18:23:30--  https://kent.dl.sourceforge.net/project/libpng/libpng16/older-releases/1.6.2/libpng-1.6.2.tar.gz
Resolving kent.dl.sourceforge.net (kent.dl.sourceforge.net)... 212.219.56.185
Connecting to kent.dl.sourceforge.net (kent.dl.sourceforge.net)|212.219.56.185|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1256016 (1,2M) [application/x-gzip]
Saving to: ‘libpng-1.6.2.tar.gz’

libpng-1.6.2.tar.gz 100%[===================>]   1,20M   959KB/s    in 1,3s    

2018-07-04 18:23:32 (959 KB/s) - ‘libpng-1.6.2.tar.gz’ saved [1256016/1256016]

Checking for gcc...
Building static library libz.a version 1.2.3 with gcc.
Checking for unistd.h... Yes.
Checking whether to use vs[n]printf() or s[n]printf()... using vs[n]printf()
Checking for vsnprintf() in stdio.h... Yes.
Checking for return value of vsnprintf()... Yes.
Checking for errno.h... Yes.
Checking for mmap support... Yes.
gcc -fPIC -DUSE_MMAP   -c -o example.o example.c
gcc -fPIC -DUSE_MMAP   -c -o adler32.o adler32.c
gcc -fPIC -DUSE_MMAP   -c -o compress.o compress.c
gcc -fPIC -DUSE_MMAP   -c -o crc32.o crc32.c
gcc -fPIC -DUSE_MMAP   -c -o gzio.o gzio.c
gcc -fPIC -DUSE_MMAP   -c -o uncompr.o uncompr.c
gcc -fPIC -DUSE_MMAP   -c -o deflate.o deflate.c
gcc -fPIC -DUSE_MMAP   -c -o trees.o trees.c
gcc -fPIC -DUSE_MMAP   -c -o zutil.o zutil.c
gcc -fPIC -DUSE_MMAP   -c -o inflate.o inflate.c
gcc -fPIC -DUSE_MMAP   -c -o infback.o infback.c
gcc -fPIC -DUSE_MMAP   -c -o inftrees.o inftrees.c
gcc -fPIC -DUSE_MMAP   -c -o inffast.o inffast.c
ar rc libz.a adler32.o compress.o crc32.o gzio.o uncompr.o deflate.o trees.o zutil.o inflate.o infback.o inftrees.o inffast.o 
gcc -fPIC -DUSE_MMAP -o example example.o -L. libz.a
gcc -fPIC -DUSE_MMAP   -c -o minigzip.o minigzip.c
gcc -fPIC -DUSE_MMAP -o minigzip minigzip.o -L. libz.a
cp zlib.h zconf.h /home/hans/local/stata-png-fix/include
chmod 644 /home/hans/local/stata-png-fix/include/zlib.h /home/hans/local/stata-png-fix/include/zconf.h
cp libz.a /home/hans/local/stata-png-fix/lib
cd /home/hans/local/stata-png-fix/lib; chmod 755 libz.a
cd /home/hans/local/stata-png-fix/lib; if test -f libz.so.1.2.3; then \
  rm -f libz.so libz.so.1; \
  ln -s libz.so.1.2.3 libz.so; \
  ln -s libz.so.1.2.3 libz.so.1; \
  (ldconfig || true)  >/dev/null 2>&1; \
fi
cp zlib.3 /home/hans/local/stata-png-fix/share/man/man3
chmod 644 /home/hans/local/stata-png-fix/share/man/man3/zlib.3
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for style of include used by make... GNU
checking dependency style of gcc... gcc3
checking dependency style of gcc... gcc3
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking for a sed that does not truncate output... /bin/sed
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for fgrep... /bin/grep -F
checking how to print strings... printf
checking for ld used by gcc... /usr/bin/x86_64-linux-gnu-ld
checking if the linker (/usr/bin/x86_64-linux-gnu-ld) is GNU ld... yes
checking how to run the C preprocessor... gcc -E
checking for gawk... (cached) gawk
checking whether ln -s works... yes
checking whether make sets $(MAKE)... (cached) yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking the maximum length of command line arguments... 1572864
checking whether the shell understands some XSI constructs... yes
checking whether the shell understands "+="... yes
checking how to convert x86_64-unknown-linux-gnu file names to x86_64-unknown-linux-gnu format... func_convert_file_noop
checking how to convert x86_64-unknown-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /usr/bin/x86_64-linux-gnu-ld option to reload object files... -r
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... ar
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for mt... mt
checking if mt is a manifest tool... no
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... yes
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/usr/bin/x86_64-linux-gnu-ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking that AWK works... ok
checking for ANSI C header files... (cached) yes
checking for an ANSI C-conforming const... yes
checking for size_t... yes
checking whether struct tm is in sys/time.h or time.h... time.h
checking for C/C++ restrict keyword... __restrict
checking for working strtod... yes
checking for memset... yes
checking for pow... no
checking for pow in -lm... yes
checking for zlibVersion in -lz... yes
checking for feenableexcept in -lm... yes
checking for feenableexcept... yes
checking if using Solaris linker... no
checking if libraries can be versioned... yes
checking for symbol prefix... 
configure: pkgconfig directory is ${libdir}/pkgconfig
configure: Extra options for compiler: 
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating libpng.pc
config.status: creating libpng-config
config.status: creating config.h
config.status: executing depfiles commands
config.status: executing libtool commands
compile-stata-png-fix.sh: line 22: stata-png.patch: No such file or directory
make  all-am
make[1]: Entering directory '/tmp/tmp.uRB1t4X5YZ/libpng-1.6.2'
rm -f pnglibconf.dfn pnglibconf.tf[45]
gawk -f ./scripts/options.awk out=pnglibconf.tf4 version=search\
    ./pngconf.h ./scripts/pnglibconf.dfa\
    ./pngusr.dfa  1>&2
gawk -f ./scripts/options.awk out=pnglibconf.tf5 pnglibconf.tf4 1>&2
rm pnglibconf.tf4
mv pnglibconf.tf5 pnglibconf.dfn
rm -f pnglibconf.out pnglibconf.c pnglibconf.tf[12]
test -d scripts || mkdir scripts || test -d scripts
echo '#include "pnglibconf.dfn"' >pnglibconf.c
gcc -E -DHAVE_CONFIG_H -I. \
     -DPNGLIB_LIBNAME='PNG16_0' -DPNGLIB_VERSION='1.6.2' -DSYMBOL_PREFIX='' -DPNG_NO_USE_READ_MACROS -DPNG_BUILDING_SYMBOL_TABLE  pnglibconf.c > pnglibconf.tf1
In file included from pnglibconf.c:1:0:
pnglibconf.dfn:33:11: fatal error: zlib.h: No such file or directory
 # include <zlib.h>
           ^~~~~~~~
compilation terminated.
Makefile:1965: recipe for target 'pnglibconf.out' failed
make[1]: *** [pnglibconf.out] Error 1
make[1]: Leaving directory '/tmp/tmp.uRB1t4X5YZ/libpng-1.6.2'
Makefile:681: recipe for target 'all' failed
make: *** [all] Error 2
ln: failed to create symbolic link '/home/hans/local/bin/xstata': No such file or directory
ln: failed to create symbolic link '/home/hans/local/bin/xstata-se': No such file or directory
ln: failed to create symbolic link '/home/hans/local/bin/xstata-mp': No such file or directory
ln: failed to create symbolic link '/home/hans/local/bin/xstata-sm': No such file or directory
```

**Comment by kylebarron** [comment]
I'm not exactly sure.

It might help to run each line of the bash script one at a time, and then see where the error is. It seems as if maybe zlib isn't building correctly?

**Comment by kylebarron** [comment]
You should run `sudo apt install build-essential` as well

**Comment by tartifola** [comment]
Hi,
thank you for providing the script. I'm trying to use it in a fresh new Linux Mint 19 environment. I have two issues:

1. line 22 of 'compile-stata-png-fix.sh' generates

          ./compile-stata-png-fix.sh: line 22: stata-png.patch: No such file or directory

    It seems easy to fix;

2. after running the script Stata still does not work; there are no question marks instead of icons  anymore but I get a "Cannot find Stata directory" and then Stata crashes.

I double checked the path to the directory and to be sure I also tried without modifing your script and put Stata directory in ~/local. I'm probably missing something stupid but I don't see it. Any help?

Thanks in advance

**Comment by kylebarron** [comment]
Actually this is probably related to #2. Can you try changing line 22 of `compile-stata-png-fix.sh` to
```
patch -p1 < ${cwd}/stata-png.patch 
```

**Comment by kylebarron** [comment]
It seems that zlib isn't linking properly for you, so that might fix it.

**Comment by kylebarron** [comment]
Try again with master

**Comment by staxpost** [comment]
Sorry for my late response, I was (still am) on vacation. I found out that after I installed the essentials (sudo apt install build-essential) the Script by Daniel Bela (https://github.com/dirtyhawk/stata-integration) worked perfectly, while your own did not. I was not sure if installing the missing essentials was enough to solve this or I did chance anything else. I did not have the time to test this thoroughly in the VM. After my return I will check this in detail and post my updates. Thanks for your help!

**Comment by staxpost** [comment]
Sorry for my late response, I was (still am) on vacation. I found out that after I installed the essentials (sudo apt install build-essential) the Script by Daniel Bela (https://github.com/dirtyhawk/stata-integration) worked perfectly, while your own did not. I was not sure if installing the missing essentials was enough to solve this or I did chance anything else. I did not have the time to test this thoroughly in the VM. After my return I will check this in detail and post my updates. Thanks for your help!
